crypten/nn/module.py

```diff
@@ -54,29 +54,71 @@ def train(self, mode=True):
         self.training = mode
 
         # Recursively set train mode
-        for module in self.modules():
+        for module in self.children():
             module.train(mode=mode)
         return self
 
     def eval(self):
         """Sets the module in evaluation mode."""
         return self.train(False)
 
+    def children(self):
+        """Returns an iterator over immediate children modules.
+        Yields:
+            Module: a child module
+        """
+        for _, module in self.named_children():
+            yield module
+
+    def named_children(self):
+        """Returns an iterator over immediate children modules, yielding both
+        the name of the module as well as the module itself.
+        Yields:
+            (string, Module): Tuple containing a name and child module
+        Example::
+            >>> for name, module in model.named_children():
+            >>>     if name in ['conv4', 'conv5']:
+            >>>         print(module)
+        """
+        memo = set()
+        for name, module in self._modules.items():
+            if module is not None and module not in memo:
+                memo.add(module)
+                yield name, module
+
     def register_module(self, name, module):
         """Registers child module in the module."""
         self._modules[name] = module
 
     def modules(self):
-        """Returns iterator over modules (non-recursively)."""
-        # TODO: Add option to do this recursively, akin to PyTorch.
+        """Returns an iterator over all modules in the network.
+        Yields:
+            Module: a module in the network
+        Note:
+            Duplicate modules are returned only once.
+        """
         for _, module in self.named_modules():
             yield module
 
-    def named_modules(self):
-        """Returns iterator over named modules (non-recursively)."""
-        # TODO: Add option to do this recursively, akin to PyTorch.
-        for name, module in self._modules.items():
-            yield name, module
+    def named_modules(self, memo=None, prefix=''):
```

Comment:
It seems a bit odd to expose `memo` (`memory` may be a better name?)  to the user this way. My suggestion would be to not expose it here, but have a separate function `_named_modules(self, memo=None, prefix="")` that you call from `named_modules`.

Also, please note we use double quotes for strings.

Reply:
I adapted this signature from pytorch for consistency. (https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=named_modules#torch.nn.Module.named_modules) But if you think having a separate function that implicitly passes memo makes more sense, I can make the corresponding change.

Reply:
Oh I didn't realize this was the PyTorch signature. Then let's keep it as you have it now!